v3/src/components/graph/adornments/adornments.tsx

```diff
@@ -9,75 +9,136 @@ import { getAdornmentContentInfo } from "./adornment-content-info"
 import { IAdornmentModel } from "./adornment-models"
 import { useInstanceIdContext } from "../../../hooks/use-instance-id-context"
 import { useTileModelContext } from "../../../hooks/use-tile-model-context"
+import { useDataConfigurationContext } from "../hooks/use-data-configuration-context"
 
 import "./adornments.scss"
 
 export const Adornments = observer(function Adornments() {
   const graphModel = useGraphModelContext(),
+    dataConfig = useDataConfigurationContext(),
     instanceId = useInstanceIdContext(),
     layout = useGraphLayoutContext(),
     { isTileSelected } = useTileModelContext(),
     adornments = graphModel.adornments
 
   if (!adornments?.length) return null
 
-  // When a graph contains multiple sub-plots, each adornment needs to be rendered once
-  // per sub-plot. We create a CSS grid in which to place the adornments, with each cell
-  // of the grid corresponding to one sub-plot.
-  const
-    xCategories = graphModel.config.categoryArrayForAttrRole("topSplit", []),
-    yCategories = graphModel.config.categoryArrayForAttrRole("rightSplit", []),
-    colCount = xCategories.length,
-    rowCount = yCategories.length,
-    { left, top, width, height } = layout.computedBounds.plot,
-    cellsRequired = Math.max(1, colCount) * Math.max(1, rowCount),
-    gridTemplateColumns = `repeat(${colCount}, 1fr)`,
-    gridTemplateRows = `repeat(${rowCount}, 1fr)`,
-    gridStyle = {
-      gridTemplateColumns,
-      gridTemplateRows,
-      height,
-      left,
-      top,
-      width,
-    }
-  let cellsRendered = 0
+  // The subPlotKey is an object that contains the attribute IDs and categorical values for the
+  // current subplot. It's used to uniquely identify the current subplot. Since it's possible to
+  // have the same attribute on two axes or splits, we need to make sure the subPlotKey is unique.
+  // So if an attribute is on more than one axis or split, we set the value of that attribute's ID
+  // to "__IMPOSSIBLE__" instead of overwriting the key's value because it's impossible for a
+  // single case to have two different values for the same attribute.
+  const updateSubPlotKey = (subPlotKey: Record<string, string>, attrId: string, cat: string) => {
+    const newSubPlotKey = { ...subPlotKey }
+    const propertyAlreadyPresent = Object.keys(newSubPlotKey).includes(attrId)
+    newSubPlotKey[attrId] = propertyAlreadyPresent && newSubPlotKey[attrId] !== cat
+      ? "__IMPOSSIBLE__"
+      : cat
```

Comment:
```suggestion
    if (cat) {
      const propertyAlreadyPresent = Object.keys(newSubPlotKey).includes(attrId)
      newSubPlotKey[attrId] = propertyAlreadyPresent && newSubPlotKey[attrId] !== cat
        ? "__IMPOSSIBLE__"
        : cat
    }
```
Vis-a-vis the discussion about handling empty strings for numeric attributes, I think if we just don't include empty strings in the `subPlotKey`, then we wouldn't need to deal with them down the road. Empty strings as categories should only occur when there are no other categories.

Reply:
@kswenson Yeah, that works -- thanks! I also had to make a small change to `setSubPlotKey` in adornment-models.ts in [the latest commit](https://github.com/concord-consortium/codap/pull/857/commits/d682bf2879c6ea37f0858d7dae0a721506aea5a6). I think the checks on 76 and 77 weren't quite right before. When I made the suggested changes in adornments.tsx and data-configuration-model.ts, the Point and Line wouldn't render correctly until I made that change to `setSubPlotKey`.